expected but , found

## Configuration (required) ##

```
version = 3.7.17
runner.dialect = scala3
```

## Command-line parameters (required) ##

When I run scalafmt via CLI like this: `scalafmt -c .scalafmt.conf Request.scala`

## Steps

Given code like this:
```scala
final case class Request(value: Option[Int], b: String)

Request(
  value = true match
    case true => Some(1)
    case _ => None,
  b = "xxx"
)
```

## Problem

Scalafmt formats code like this:
```scala
Request.scala:6: error: [dialect scala3] ; expected but , found
    case _ => None,
```

## Expectation

I would like the formatted output to look like this:
```scala
final case class Request(value: Option[Int], b: String)

Request(
  value = true match
    case true => Some(1)
    case _ => None,
  b = "xxx"
)
```

## Workaround

Exclude files like this.

--- a/scalameta/parsers/shared/src/main/scala/scala/meta/internal/parsers/LazyTokenIterator.scala
+++ b/scalameta/parsers/shared/src/main/scala/scala/meta/internal/parsers/LazyTokenIterator.scala
@@ -22,8 +22,7 @@ private[parsers] class LazyTokenIterator private (
   import scannerTokens._
 
   private def getNextTokenRef(): TokenRef = {
-    if (curr.next eq null) curr.next = nextToken(curr)
-    curr.next
+    if (curr.next eq null) nextToken(curr) else curr.next
   }
 
   override def next(): Unit = {

--- a/scalameta/parsers/shared/src/main/scala/scala/meta/internal/parsers/ScannerTokens.scala
+++ b/scalameta/parsers/shared/src/main/scala/scala/meta/internal/parsers/ScannerTokens.scala
@@ -475,16 +475,42 @@ final class ScannerTokens(val tokens: Tokens)(implicit dialect: Dialect) {
     def nonTrivial(sepRegions: List[SepRegion]) = curr match {
       case _: EOF => getAtEof(sepRegions)
       case _: Comma =>
-        if (inParens(sepRegions))
+        if (inParens(sepRegions)) {
+          @tailrec
+          def findLastAndPrevRef(ref: TokenRef, prev: TokenRef): (TokenRef, TokenRef) = {
+            val next = ref.next
+            if (next eq null) (ref, prev) else findLastAndPrevRef(next, ref)
+          }
+          def swapWithOutdents(tokenRef: TokenRef) = {
+            @tailrec
+            def iter(oref: TokenRef): TokenRef = {
+              val nref = nextToken(oref)
+              if (nref.token.is[Indentation.Outdent]) iter(nref)
+              else if (oref eq tokenRef) tokenRef
+              else {
+                oref.next = tokenRef
+                try tokenRef.next
+                finally tokenRef.next = nref
+              }
+            }
+            iter(tokenRef)
+          }
           mkOutdentsOpt(currPos, sepRegions) {
             case (r: SepRegionIndented) :: rs => (Left(r), rs)
             case (_: RegionNonDelimNonIndented) :: rs => (Right(true), rs)
             case rs => (Right(false), rs)
           } match {
-            case Right(tokenRef) => tokenRef
-            case Left(regions) => currRef(regions)
+            case Right(tokenRef) =>
+              val (last, prev) = findLastAndPrevRef(tokenRef, null)
+              val cref = swapWithOutdents(last)
+              if (prev eq null) cref
+              else {
+                if (cref ne last) prev.next = cref
+                tokenRef
+              }
+            case Left(regions) => swapWithOutdents(currRef(regions))
           }
-        else currRef(sepRegions)
+        } else currRef(sepRegions)
       case _: KwEnum => currRef(RegionTemplateMark :: sepRegions)
       case _: KwObject | _: KwClass | _: KwTrait | _: KwPackage | _: KwNew
           if dialect.allowSignificantIndentation =>

--- a/scalameta/parsers/shared/src/main/scala/scala/meta/internal/parsers/ScannerTokens.scala
+++ b/scalameta/parsers/shared/src/main/scala/scala/meta/internal/parsers/ScannerTokens.scala
@@ -346,8 +346,11 @@ final class ScannerTokens(val tokens: Tokens)(implicit dialect: Dialect) {
   }
 
   @inline
-  private[parsers] def nextToken(ref: TokenRef): TokenRef =
-    nextToken(ref.token, ref.pos, ref.nextPos, ref.regions)
+  private[parsers] def nextToken(ref: TokenRef): TokenRef = {
+    val next = nextToken(ref.token, ref.pos, ref.nextPos, ref.regions)
+    ref.next = next
+    next
+  }
 
   @tailrec
   private[parsers] def nextToken(

--- a/tests/shared/src/test/scala/scala/meta/tests/parsers/dotty/SignificantIndentationSuite.scala
+++ b/tests/shared/src/test/scala/scala/meta/tests/parsers/dotty/SignificantIndentationSuite.scala
@@ -1641,11 +1641,32 @@ class SignificantIndentationSuite extends BaseDottySuite {
          |  b = "xxx"
          |)
          |""".stripMargin
-    val error =
-      """|<input>:4: error: ; expected but , found
-         |    case _ => None,
-         |                  ^""".stripMargin
-    runTestError[Stat](code, error)
+    val layout =
+      """|Request(value = true match {
+         |  case true =>
+         |    Some(1)
+         |  case _ =>
+         |    None
+         |}, b = "xxx")
+         |""".stripMargin
+    val tree = Term.Apply(
+      tname("Request"),
+      List(
+        Term.Assign(
+          tname("value"),
+          Term.Match(
+            bool(true),
+            List(
+              Case(bool(true), None, Term.Apply(tname("Some"), List(int(1)))),
+              Case(Pat.Wildcard(), None, tname("None"))
+            ),
+            Nil
+          )
+        ),
+        Term.Assign(tname("b"), str("xxx"))
+      )
+    )
+    runTestAssert[Stat](code, layout)(tree)
   }
 
   test("indented-double-apply") {

--- a/tests/shared/src/test/scala/scala/meta/tests/parsers/dotty/SignificantIndentationSuite.scala
+++ b/tests/shared/src/test/scala/scala/meta/tests/parsers/dotty/SignificantIndentationSuite.scala
@@ -1632,6 +1632,22 @@ class SignificantIndentationSuite extends BaseDottySuite {
     )
   }
 
+  test("#3531 apply with optional braces and trailing comma") {
+    val code =
+      """|Request(
+         |  value = true match
+         |    case true => Some(1)
+         |    case _ => None,
+         |  b = "xxx"
+         |)
+         |""".stripMargin
+    val error =
+      """|<input>:4: error: ; expected but , found
+         |    case _ => None,
+         |                  ^""".stripMargin
+    runTestError[Stat](code, error)
+  }
+
   test("indented-double-apply") {
     runTestAssert[Stat](
       """|def method = 

